### PR TITLE
all: fix setupEmbedCfg functions to set cfg.InitialCluster more safely

### PR DESCRIPTION
**Pull request by callthingsoff**
When the slice is empty, it can panic with "index 1 out of bounds". This patch rewrites these functions to set cfg.InitialCluster more safely.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/integration/embed/embed_test.go [hunk 1]
```diff
@@ -206,11 +206,15 @@ func setupEmbedCfg(cfg *embed.Config, curls []url.URL, purls []url.URL) {
...
 	cfg.ListenClientUrls, cfg.AdvertiseClientUrls = curls, curls
 	cfg.ListenPeerUrls, cfg.AdvertisePeerUrls = purls, purls
-	cfg.InitialCluster = ""
-	for i := range purls {
-		cfg.InitialCluster += ",default=" + purls[i].String()
+
+	if len(purls) > 0 {
+		builder := new(strings.Builder)
+		builder.WriteString(fmt.Sprintf("default=%s", purls[0].String()))
+		for _, purl := range purls[1:] {
```
Comment (siyuanfoundation) [inline tests/integration/embed/embed_test.go:213]: how about using `"default=" + strings.Join(purls, ",default=")`

Comment (callthingsoff) [reply tests/integration/embed/embed_test.go:213]: `purls []url.URL` is not `[]string`
```go
go doc strings.Join

package strings // import "strings"

func Join(elems []string, sep string) string
    Join concatenates the elements of its first argument to create a single
    string. The separator string sep is placed between elements in the resulting
    string.
```

Comment (siyuanfoundation) [reply tests/integration/embed/embed_test.go:213]: oh, I see

File: tests/integration/embed/embed_test.go [hunk 2]
```diff
@@ -206,11 +206,15 @@ func setupEmbedCfg(cfg *embed.Config, curls []url.URL, purls []url.URL) {
 	cfg.ClusterState = "new"
 	cfg.ListenClientUrls, cfg.AdvertiseClientUrls = curls, curls
 	cfg.ListenPeerUrls, cfg.AdvertisePeerUrls = purls, purls
-	cfg.InitialCluster = ""
-	for i := range purls {
-		cfg.InitialCluster += ",default=" + purls[i].String()
+
```
Comment (siyuanfoundation) [inline tests/integration/embed/embed_test.go:209]:
```suggestion
builder := new(strings.Builder)
for i, purl := range(purls) {
    if i == 0 {
        builder.WriteString(fmt.Sprintf("default=%s",purl.String()))
    } else {
        builder.WriteString(fmt.Sprintf(",default=%s",purl.String()))
    }
}
cfg.InitialCluster = builder.String()
```

Comment (callthingsoff) [reply tests/integration/embed/embed_test.go:209]: Thanks, personally I prefer the previous version of mine.
It's consistent with other code and faster.

**Comment by callthingsoff** [comment]
cc @ahrtr @serathius @jmhbnz

File: tests/integration/embed/embed_test.go [hunk 3]
```diff
@@ -206,11 +206,15 @@ func setupEmbedCfg(cfg *embed.Config, curls []url.URL, purls []url.URL) {
...
+	if len(purls) > 0 {
+		builder := new(strings.Builder)
+		builder.WriteString(fmt.Sprintf("default=%s", purls[0].String()))
```
Comment (ivanvc) [inline tests/integration/embed/embed_test.go:212]: I don't think the `url.URL` needs explicit casting into a string. The following should have the same result.

```golang
fmt.Sprintf("default=%s", purls[0])
```

File: tools/etcd-dump-metrics/etcd.go [hunk 4]
```diff
@@ -53,11 +53,14 @@ func setupEmbedCfg(cfg *embed.Config, curls, purls, ics []url.URL) {
 	cfg.ListenClientUrls, cfg.AdvertiseClientUrls = curls, curls
 	cfg.ListenPeerUrls, cfg.AdvertisePeerUrls = purls, purls
 
-	cfg.InitialCluster = ""
-	for i := range ics {
-		cfg.InitialCluster += fmt.Sprintf(",%d=%s", i, ics[i].String())
+	if len(ics) > 0 {
+		builder := new(strings.Builder)
+		builder.WriteString(fmt.Sprintf("%d=%s", 0, ics[0].String()))
```
Comment (ivanvc) [inline tools/etcd-dump-metrics/etcd.go:58]: See my previous comment. And, in this case, it could also be simplified to:

```go
fmt.Sprintf("0=%s", ics[0])
```

File: tests/integration/embed/embed_test.go [hunk 5]
```diff
@@ -206,11 +206,15 @@ func setupEmbedCfg(cfg *embed.Config, curls []url.URL, purls []url.URL) {
...
+		for _, purl := range purls[1:] {
+			builder.WriteString(fmt.Sprintf(",default=%s", purl.String()))
```
Comment (ivanvc) [inline tests/integration/embed/embed_test.go:214]: Ditto

File: tools/etcd-dump-metrics/etcd.go [hunk 6]
```diff
@@ -53,11 +53,14 @@ func setupEmbedCfg(cfg *embed.Config, curls, purls, ics []url.URL) {
...
+		for i, ic := range ics[1:] {
+			builder.WriteString(fmt.Sprintf(",%d=%s", i, ic.String()))
```
Comment (ivanvc) [inline tools/etcd-dump-metrics/etcd.go:60]: Ditto.

**Review by ivanvc — COMMENTED** [review]
Thanks for your PR @callthingsoff. I left a possible improvement for your code :v:

**Comment by serathius** [comment]
QQ, if you are touching this code. Have you double checked if it has any tests? If not please start from adding tests.

File: tests/integration/embed/embed_test.go (hunk 3 above)
Comment (callthingsoff) [reply tests/integration/embed/embed_test.go:212]:
```go
fmt.Sprintf("default=%s", &purls[0]
```
I think you mean `&purls[0]`
```go
go doc url.URL.String
package url // import "net/url"

func (u *URL) String() string
```
Thanks, but I think it's just fine to call String() here and below.

File: tools/etcd-dump-metrics/etcd.go (hunk 4 above)
Comment (callthingsoff) [reply tools/etcd-dump-metrics/etcd.go:58]:
```go
fmt.Sprintf("default=%s", &purls[0]
```
See my comment above.

File: tests/integration/embed/embed_test.go (hunk 5 above)
Comment (callthingsoff) [reply tests/integration/embed/embed_test.go:214]: See my comment above.

File: tools/etcd-dump-metrics/etcd.go (hunk 6 above)
Comment (callthingsoff) [reply tools/etcd-dump-metrics/etcd.go:60]: See my comment above.

File: tests/integration/embed/embed_test.go (hunk 3 above)
Comment (ivanvc) [reply tests/integration/embed/embed_test.go:216]: Good point; I didn't check the type of `purls`.

File: tests/integration/embed/join_urls_test.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+			urls: []url.URL{
+				parseURL("http://example.com/#x"),
+				parseURL("http://example.com/#y"),
+				parseURL("http://example.com/#z"),
+			},
+			want: "default=http://example.com/#x,default=http://example.com/#y,default=http://example.com/#z",
+		},
+	}
+	for i, tt := range tests {
+		if got := joinURLs(tt.urls); got != tt.want {
```
Comment (ahrtr) [inline tests/integration/embed/join_urls_test.go:62]: `joinURLs` is already part of the test, we are adding test for test. Not necessary to me.

File: tools/etcd-dump-metrics/etcd.go [hunk 8]
```diff
@@ -53,11 +53,15 @@ func setupEmbedCfg(cfg *embed.Config, curls, purls, ics []url.URL) {
 	cfg.ListenClientUrls, cfg.AdvertiseClientUrls = curls, curls
 	cfg.ListenPeerUrls, cfg.AdvertisePeerUrls = purls, purls
 
-	cfg.InitialCluster = ""
-	for i := range ics {
-		cfg.InitialCluster += fmt.Sprintf(",%d=%s", i, ics[i].String())
```
Comment (ahrtr) [inline tools/etcd-dump-metrics/etcd.go:58]: It seems an unnecessary change,
- The function `setupEmbedCfg` is already very simple. 
- The new function `joinURLs` isn't used by multiple users (it has only one user).

**Comment by callthingsoff** [comment]
Thanks, closing this.
